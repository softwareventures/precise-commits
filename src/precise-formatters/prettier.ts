import * as fs from "fs";
import {extname, join} from "path";
import ignore from "ignore";
import type {Options as PrettierOptions} from "prettier";
import {getSupportInfo, format, resolveConfig, check} from "prettier";
import type {PreciseFormatter} from "../precise-formatter";
import type {CharacterRange} from "../utils";

let prettierSupportedFileExtensions: string[] = [];
getSupportInfo().languages.forEach(language => {
    prettierSupportedFileExtensions = [
        ...prettierSupportedFileExtensions,
        ...(language.extensions ?? [])
    ];
});

export const preciseFormatterPrettier: PreciseFormatter<PrettierOptions> = {
    /**
     * Resolve the relevant prettier config for the given
     * modified file path.
     */
    resolveConfig: async (modifiedFilePath: string) =>
        resolveConfig(modifiedFilePath, {useCache: false}).then(config => ({
            ...config,
            filepath: modifiedFilePath
        })),
    /**
     * Return true if the whole file has already been formatted appropriately based on
     * the resolved prettier config. We can use this as a check to skip unnecessary work.
     */
    isAlreadyFormatted: async (fileContents: string, config: PrettierOptions | null) =>
        check(fileContents, {...config}),
    /**
     * Run prettier's check mode on the given ranges and return true if they are all
     * already formatted appropriately based on the given prettier config.
     */
    checkFormattingOfRanges: async (
        filePath: string,
        fileContents: string,
        config: PrettierOptions | null,
        characterRanges: CharacterRange[]
    ) => {
        const formattedContents = fileContents;
        return characterRanges.every(characterRange =>
            check(formattedContents, {
                ...config,
                filepath: filePath,
                rangeStart: characterRange.rangeStart,
                rangeEnd: characterRange.rangeEnd
            })
        );
    },
    /**
     * Run prettier on each character range pair given, and apply the
     * difference as a patch to the original contents using an implementation
     * of the Myer's diff algorithm.
     */
    formatRanges: async (
        filePath: string,
        fileContents: string,
        config: PrettierOptions | null,
        characterRanges: CharacterRange[]
    ) =>
        // Start from the last character range and work backwards so that
        // we don't have to update character ranges to account for the changes
        // we've already made.
        characterRanges.reduceRight(
            (fileContents, {rangeStart, rangeEnd}) =>
                format(fileContents, {
                    ...config,
                    filepath: filePath,
                    rangeStart,
                    rangeEnd
                }),
            fileContents
        ),
    /**
     * Generate a predicate function which will return true if the filename
     * is not excluded via a .prettierignore file.
     */
    generateIgnoreFilePredicate: (workingDirectory: string) => {
        const prettierIgnoreFilePath = join(workingDirectory, ".prettierignore");
        /**
         * If there is no .prettierignore file present, simply always return true
         * from the predicate
         */
        if (!fs.existsSync(prettierIgnoreFilePath)) {
            return () => true;
        }
        /**
         * Use "ignore"'s createFilter() method to create a predicate
         */
        const prettierIgnoreFileContents = fs.readFileSync(prettierIgnoreFilePath, "utf8");
        return ignore().add(prettierIgnoreFileContents).createFilter();
    },
    /**
     * Return true if prettier supports the file extension of the given
     * filename.
     */
    hasSupportedFileExtension: (filename: string) => {
        const fileExtension = extname(filename);
        return prettierSupportedFileExtensions.includes(fileExtension);
    }
};
