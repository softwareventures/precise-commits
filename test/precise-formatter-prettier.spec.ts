import {readFixtures} from "./test-utils";
import {preciseFormatterPrettier} from "../src/precise-formatters/prettier";

const fixtures = readFixtures();

describe("preciseFormatterPrettier", () => {
    describe("resolveConfig()", () => {
        fixtures.forEach(({fixtureName, fileExtension}) => {
            it(fixtureName, () => {
                const prettierConfig = preciseFormatterPrettier.resolveConfig(
                    `./test/fixtures/${fixtureName}/initial${fileExtension}`
                );
                expect(prettierConfig).toMatchSnapshot();
            });
        });
    });

    describe("formatRanges()", () => {
        it("should format the given ranges of the given source", () => {
            const contents = `
        var a = 1
        var b = 2
        var c = 3
      `;
            const formatted = preciseFormatterPrettier.formatRanges(
                "foo.js",
                contents,
                {
                    semi: true
                },
                [
                    {rangeStart: 0, rangeEnd: 10},
                    {rangeStart: 44, rangeEnd: 60}
                ]
            );
            expect(formatted).toEqual(`
        var a = 1;
        var b = 2
        var c = 3;
      `);
        });
    });

    describe("checkFormattingOfRanges()", () => {
        it("should return true if the given ranges are all formatted according to the given config", () => {
            const contents = `
        var a = 1;
        var b = 2
        var c = 3;
      `;
            const formatted = preciseFormatterPrettier.checkFormattingOfRanges(
                "foo.js",
                contents,
                {
                    semi: true
                },
                [
                    {rangeStart: 0, rangeEnd: 10},
                    {rangeStart: 46, rangeEnd: 62}
                ]
            );
            expect(formatted).toEqual(true);
        });

        it("should return false if any of the given ranges are not formatted according to the given config", () => {
            const contents = `
        var a = 1
        var b = 2
        var c = 3
      `;
            const formatted = preciseFormatterPrettier.checkFormattingOfRanges(
                "foo.js",
                contents,
                {
                    semi: true
                },
                [
                    {rangeStart: 0, rangeEnd: 10},
                    {rangeStart: 46, rangeEnd: 62}
                ]
            );
            expect(formatted).toEqual(false);
        });
    });
});
