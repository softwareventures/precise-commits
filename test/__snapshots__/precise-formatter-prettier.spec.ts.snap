// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preciseFormatterPrettier resolveConfig() brand-new-json-file 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/brand-new-json-file/initial.json",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 2,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;

exports[`preciseFormatterPrettier resolveConfig() brand-new-jsx-file 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/brand-new-jsx-file/initial.jsx",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 4,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;

exports[`preciseFormatterPrettier resolveConfig() changed-line-in-closure 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/changed-line-in-closure/initial.ts",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 4,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;

exports[`preciseFormatterPrettier resolveConfig() empty-javascript-file 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/empty-javascript-file/initial.js",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 4,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;

exports[`preciseFormatterPrettier resolveConfig() formatting-would-affect-total-line-count 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/formatting-would-affect-total-line-count/initial.js",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 4,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;

exports[`preciseFormatterPrettier resolveConfig() hunk-marker-in-code 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/hunk-marker-in-code/initial.ts",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 4,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;

exports[`preciseFormatterPrettier resolveConfig() multiple-updated-lines-javascript-file 1`] = `
{
  "endOfLine": "auto",
  "filepath": "./test/fixtures/multiple-updated-lines-javascript-file/initial.js",
  "semi": true,
  "singleQuote": false,
  "trailingComma": "none",
}
`;

exports[`preciseFormatterPrettier resolveConfig() single-updated-line-typescript 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/single-updated-line-typescript/initial.ts",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 4,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;

exports[`preciseFormatterPrettier resolveConfig() updated-class-body 1`] = `
{
  "arrowParens": "avoid",
  "bracketSameLine": false,
  "bracketSpacing": false,
  "endOfLine": "auto",
  "filepath": "./test/fixtures/updated-class-body/initial.js",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "jsxSingleQuote": false,
  "printWidth": 100,
  "proseWrap": "always",
  "quoteProps": "consistent",
  "requirePragma": false,
  "semi": true,
  "singleQuote": false,
  "tabWidth": 4,
  "trailingComma": "none",
  "useTabs": false,
  "vueIndentScriptAndStyle": true,
}
`;
